mdoc/Mono.Documentation/MDocUpdater.cs

```diff
@@ -3626,124 +3623,129 @@ static bool UpdateAssemblyVersionForAssemblyInfo (XmlElement e, XmlElement root,
-                        alreadyExists.XElement.SetAttribute (Consts.FrameworkAlternate, newfxlist);
-                    }
-                    continue;
-                }
-                else // let's create it
-                {
-                    XmlElement ae = root.OwnerDocument.CreateElement ("Attribute");
-                    e.AppendChild (ae);
-                    WriteElementText (ae, "AttributeName", currentAttribute.Key);
-
-                    if (fx.FrameworksCount > 1)
-                    {
-                        ae.SetAttribute (Consts.FrameworkAlternate, fx.Name);
-                    }
+                e = root.OwnerDocument.CreateElement("Attributes");
+
+            // we'll use C# version as key identifier of each attribute
+            var existingAttrElements = e.ChildNodes.SafeCast<XmlElement>()
+                                .ToDictionary(
+                                  elem => elem.SelectSingleNode("/AttributeName[@" + Language + "='" + Consts.CSharp + "']")?.InnerText
+                                          ?? elem.FirstChild.InnerText
+                                  , elem => elem);
+            var currentAttrBatch = new HashSet<string>();
+
+            if (customAttributesWithPrefix != null)
+            {
+                foreach (var customAttrWithPrefix in customAttributesWithPrefix)
+                {
+                    var customAttr = customAttrWithPrefix.Item1;
+                    var prefix = customAttrWithPrefix.Item2;
+                    if (FormatterManager.CSharpAttributeFormatter.TryGetAttributeString(customAttr, out string csharpAttrStr, prefix, !isAssemblyAttribute))
+                    {
+                        currentAttrBatch.Add(csharpAttrStr);
+                        if (existingAttrElements.TryGetValue(csharpAttrStr, out XmlElement elem))
+                        {
+                            if (fx.FrameworksCount > 1)
+                            {
+                                var newfxlist = FXUtils.AddFXToList(elem.GetAttribute(Consts.FrameworkAlternate), fx.Name);
+                                elem.SetAttribute(Consts.FrameworkAlternate, newfxlist);
+                            }
+                            continue;
+                        }
+                        else // let's create it
+                        {
+                            XmlElement ae = root.OwnerDocument.CreateElement("Attribute");
+                            e.AppendChild(ae);
+                            var node = WriteElementText(ae, "AttributeName", csharpAttrStr, forceNewElement: true);
+                            if (!isAssemblyAttribute)
+                            {
+                                WriteElementAttribute(node, Language, Consts.CSharp);
+
+                                foreach (var formatter in FormatterManager.AdditionalAttributeFormatters)

```

Comment:
If you end up changing it per the previous comment, I would sort of invert this, where I iterate through all the relevant formatters (C# would be the default unless you added additional ones via -lang) ... and then just use that formatter's language (I assume there's a language attribute string?) to query for whether the attribute element exists.